### PR TITLE
Enable permissive provenance by default

### PR DESCRIPTION
**Pull request by RalfJung**
This completes the plan laid out in https://github.com/rust-lang/miri/issues/2133:
- We use permissive provenance with wildcard pointers by default.
- We print a warning on int2ptr casts. `-Zmiri-permissive-provenance` suppresses the warning; `-Zmiri-strict-provenance` turns it into a hard error.
- Raw pointer tagging is now always enabled, so we remove the `-Zmiri-tag-raw-pointers` flag and the code for untagged pointers. (Passing the flag still works, for compatibility -- but we just ignore it, with a warning.)

We also fix an intptrcast issue:
- Only live allocations are considered when computing the AllocId from an address.

So, finally, Miri has a good story for ptr2int2ptr roundtrips *and* no weird false negatives when doing raw pointer stuff with Stacked Borrows. :-) :tada:   Thanks a lot to everyone who helped with this, in particular @carbotaniuman who convinced me this is even possible.

Fixes https://github.com/rust-lang/miri/issues/2133
Fixes https://github.com/rust-lang/miri/issues/1866
Fixes https://github.com/rust-lang/miri/issues/1993

### CODE REVIEW

File: src/machine.rs [hunk 1]
```diff
@@ -140,8 +140,9 @@ pub enum Tag {
 
 #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
 static_assert_size!(Pointer<Tag>, 24);
+// FIXME: this would with in 24bytes but layout optimizations are not smart enough
 // #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
-// static_assert_size!(Pointer<Option<Tag>>, 24);
+//static_assert_size!(Pointer<Option<Tag>>, 24);
```
Comment (RalfJung) [inline src/machine.rs:145]: I am quite bummed by this... `Tag` now has two `NonZeroU64`, so we could use one to indicate `Wildcard` and one to indicate `None`, but it seems rustc is not smart enough for that. :(

I wonder if it's worth implementing the packing by hand, or if having the extra 8 bytes is just not such a big deal...

File: src/stacked_borrows/diagnostics.rs [hunk 2]
```diff
@@ -125,74 +116,17 @@ impl AllocHistory {
...
-                            None
-                        }
-                    })
-            });
-
-            Some(TagHistory::Untagged {
-                recently_created,
-                matching_created,
-                recently_invalidated,
-                protected,
```
Comment (RalfJung) [inline src/stacked_borrows/diagnostics.rs:193]: @saethlin a lot of your code turned into dead code here...

Comment (saethlin) [reply src/stacked_borrows/diagnostics.rs:193]: That's probably a good thing. This code implements a bunch of best-effort heuristics which are hard to explain to a new user.

I haven't looked into the state of the new diagnostics yet. I wouldn't be surprised if we need to add in a new branch somewhere for permissive provenance, but this code going away on its own does not concern me.

**Review by oli-obk — APPROVED** [review]
I'm a bit surprised about all the behaviour changes. I would have expected this to just change the defaults. I guess the subsequent cleanups are the cause here?

I wish we had a good way to run perf. Maybe we could figure out a way to run a separate perfbot suite via the same infrastructure... I'll check that out

**Comment by oli-obk** [comment]
> The remaining output changes come from the fact that we now enable raw pointer tagging by default. 

yea I got that part.

the rest seems ok to include here, I was just surprised, as the PR title and description did not make me think of anything beyond the test changes due to changes in defaults.

**Comment by RalfJung** [comment]
That test failure is very strange, I cannot reproduce it locally
```
Testing `cargo miri test` (no isolation, no doctests)...
Test stdout or stderr did not match reference!
--- BEGIN test stdout ---
running 2 tests
..
test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
imported main
running 8 tests
..i--- END test stdout ---
--- BEGIN test stderr ---
error: Undefined Behavior: 0x93e046 is not a valid pointer
--> /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/macos.rs:24:32
|
24|            let ret = unsafe { func(chunk.as_mut_ptr(), chunk.len()) };
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^0x93e046 is not a valid pointer
|
= help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
= help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
= note: inside `getrandom::imp::getrandom_inner` at /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/macos.rs:24:32
= note: inside `getrandom::getrandom` at /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/lib.rs:235:5
= note: inside `<rand::prelude::SmallRng as rand::SeedableRng>::from_entropy` at /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/rand_core-0.6.2/src/lib.rs:380:27
note: inside `entropy_rng` at tests/test.rs:33:19
--> tests/test.rs:33:19
|
33|    let mut rng = SmallRng::from_entropy();
| ^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by RalfJung** [comment]
The function pointer lacks provenance. Maybe I didn't think this part about adjacent allocations through carefully enough... I think I'll back that part out again.

**Comment by RalfJung** [comment]
Ah bummer, getrandom 1 still needs permissive provenance on macOS...

**Comment by saethlin** [comment]
When running `cargo miri test`, I think the ptr-int cast warnings are eaten by the test runner for lib/bin tests. For example, running the tests of `string_cache`:
```
     Running tests/small-stack.rs (target/miri/x86_64-unknown-linux-gnu/debug/deps/small_stack-936c74f9283f43cc)
error: Undefined Behavior: trying to reborrow <wildcard> for SharedReadOnly permission at alloc4271[0x0], but no exposed tags have suitable permission in the borrow stack for this location
   --> /root/build/src/atom.rs:233:25
    |
233 |             if unsafe { &*entry }.ref_count.fetch_sub(1, SeqCst) == 1 {
    |                         ^^^^^^^
    |                         |
    |                         trying to reborrow <wildcard> for SharedReadOnly permission at alloc4271[0x0], but no exposed tags have suitable permission in the borrow stack for this location
    |                         this error occurs as part of a reborrow at alloc4271[0x0..0x8]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
            
    = note: inside `<string_cache::Atom<string_cache::EmptyStaticAtomSet> as std::ops::Drop>::drop` at /root/build/src/atom.rs:233:25
    = note: inside `std::ptr::drop_in_place::<string_cache::Atom<string_cache::EmptyStaticAtomSet>> - shim(Some(string_cache::Atom<string_cache::EmptyStaticAtomSet>))` at /root/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:487:1
note: inside closure at tests/small-stack.rs:13:9
   --> tests/small-stack.rs:13:9
    |
13  |         })
    |         ^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

   Doc-tests string_cache
```

Amusingly, one of the doctests runs into the same sort of issue, but in that case the ptr-int cast warning is displayed before the UB error.

**Comment by RalfJung** [comment]
I have seen the warning in `cargo miri test`. In fact this reproduces inside `test-cargo-miri`: `cargo miri test --target aarch64-apple-darwin` gives
```
     Running tests/test.rs (target/miri/aarch64-apple-darwin/debug/deps/test-8bd6ea9154753acd)

running 8 tests
test cargo_env ... ok
test do_panic - should panic ... ok
test does_not_work_on_miri ... ignored
test entropy_rng ... warning: pointer-to-integer cast
  --> /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.1.16/src/util_libc.rs:98:22
   |
98 |         NonNull::new(addr as *mut _)
   |                      ^^^^^^^^^^^^^^ pointer-to-integer cast
   |
   = help: this program is using integer-to-pointer casts or (equivalently) `from_exposed_addr`,
   = help: which means that Miri might miss pointer bugs in this program
   = help: see https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html for more details on that operation
   = help: to ensure that Miri does not miss bugs in your program, use `with_addr` (https://doc.rust-lang.org/nightly/std/ptr/index.html#strict-provenance) instead
   = help: you can then pass the `-Zmiri-strict-provenance` flag to Miri, to ensure you are not relying on `from_exposed_addr` semantics
   = help: alternatively, the `-Zmiri-permissive-provenance` flag disables this warning
           
   = note: inside `getrandom::util_libc::Weak::ptr` at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.1.16/src/util_libc.rs:98:22
   = note: inside `getrandom::imp::getrandom_inner` at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.1.16/src/macos.rs:18:25
   = note: inside `getrandom::getrandom` at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.1.16/src/lib.rs:291:5
note: inside `entropy_rng` at tests/test.rs:29:5
  --> tests/test.rs:29:5
   |
29 |     getrandom_1::getrandom(&mut data).unwrap();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure at tests/test.rs:26:1
  --> tests/test.rs:26:1
   |
25 |   #[test]
   |   ------- in this procedural macro expansion
26 | / fn entropy_rng() {
27 | |     // Test `getrandom` directly (in multiple different versions).
28 | |     let mut data = vec![0; 16];
29 | |     getrandom_1::getrandom(&mut data).unwrap();
...  |
42 | |     let _val = rng.gen::<i128>();
43 | | }
   | |_^
   = note: this warning originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

ok
test fail_index_check - should panic ... ok
```

File: src/intptrcast.rs [hunk 3]
```diff
@@ -120,42 +120,43 @@ impl<'mir, 'tcx> GlobalStateInner {
...
-                Pointer::new(
-                    alloc_id.map(|alloc_id| Tag::Concrete { alloc_id, sb: SbTag::Untagged }),
-                    Size::from_bytes(addr),
-                )
+            ProvenanceMode::Default => {
+                // The first time this happens, print a warning.
+                use std::sync::atomic::{AtomicBool, Ordering};
+                static FIRST_WARNING: AtomicBool = AtomicBool::new(true);
+                if FIRST_WARNING.swap(false, Ordering::Relaxed) {
+                    register_diagnostic(NonHaltingDiagnostic::Int2Ptr);
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/intptrcast.rs:147]: *oops* int2ptr is right

**Comment by RalfJung** [comment]
One thing I was wondering about: currently we only print the warning on the first cast.
We might instead keep a `HashSet` of `Span`s that we have reported and print the warning the first time it is triggered for each unique span? (possibly with some kind of limit, like stopping to warn after 10 or 100 warnings or so)

**Comment by saethlin** [comment]
That may be useful to have behind a flag.

What I am thinking of for diagnostics, is that when we encounter a problem we should be able to report what Miri knows about what offsets and with what permissions have been exposed in an allocation. I'd have to actually get to implementing this to know how useful it is, but for example in `string_cache` we have an invalid use of Wildcard for SRO at offset 0, which makes me think we just didn't expose that address. Most of the other errors I'm seeing are for SRW.

**Comment by RalfJung** [comment]
> That may be useful to have behind a flag.

So many flags.^^ For now I'd like to have a reasonable default.
And I feel reporting not only the first hit might be useful...

> for example in string_cache we have an invalid use of Wildcard for SRO at offset 0, which makes me think we just didn't expose that address. Most of the other errors I'm seeing are for SRW.

Provenances (AllocId + SbTag) are exposed, not addresses. And if the AllocId had not been exposed we would not even have gotten to the Stacked Borrows part.

But likely the tag that was exposed is already invalid, or never was valid for this location.

**Comment by saethlin** [comment]
In real code, especially from a test suite, you can expect 1-2 backtraces to fit on a screen. I think if Miri emits 10, people will rapidly decide to always silence them, just like how people learn to always disable isolation.

**Comment by saethlin** [comment]
Many of the int2ptr warnings I've seen are for `once_cell`. I should submit my patch, but also this warning is pretty far off the mark. Perhaps this is another good place to use the `is_local` logic, if you're looking to filter? A user probably cares more about the warning when it's for something they wrote, rather than a dependency.

**Comment by RalfJung** [comment]
> In real code, especially from a test suite, you can expect 1-2 backtraces to fit on a screen. I think if Miri emits 10, people will rapidly decide to always silence them, just like how people learn to always disable isolation.

If they are not interested in removing these casts, that's fine.
But if they are interested in removing them, I would be quite annoyed if Miri told me the first cast, I fix that, the I rerun and learn about the 2nd cast, fix that, rinse, repeat.

> Many of the int2ptr warnings I've seen are for once_cell. I should submit my patch, but also this warning is pretty far off the mark. Perhaps this is another good place to use the is_local logic, if you're looking to filter? A user probably cares more about the warning when it's for something they wrote, rather than a dependency.

Well, I do want to make sure people are aware that such a precision-losing cast happens -- even if they cannot do anything about it, IMO it is important that they can know that Miri might miss some bugs here. This warning is load-bearing, in a sense.

We could prune the backtrace but I doubt that would help much...

**Comment by RalfJung** [comment]
I'm going to land this for now, we can always fine-tune it later.
Thanks all for the helpful feedback. :)

@bors r+

**Pull request merged**